File: audio/src/main/kotlin/app/guitar/audio/PitchAnalysis.kt
package app.guitar.audio

import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.round

/**
 * Equal-tempered pitch math. The A4 reference is configurable (435-445 Hz is the
 * usual range) so the user can tune to anything from baroque pitch (A=415) up to
 * modern brilliant orchestras (A=445) without changing anything else.
 */
data class PitchEstimate(
    /** Detected fundamental in Hz (or the input freq). */
    val freqHz: Float,
    /** Nearest equal-tempered MIDI note, clamped to 0..127. */
    val midi: Int,
    /** Signed cents offset from the nearest note, in (-50, +50]. */
    val cents: Float,
)

object PitchAnalysis {

    /** Convert a frequency in Hz to a (nearest MIDI note, cents offset).
     *  cents is in (-50, +50]. */
    fun analyze(freqHz: Float, a4Hz: Float = 440f): PitchEstimate {
        require(freqHz > 0f) { "freq must be > 0, got $freqHz" }
        require(a4Hz > 0f) { "A4 reference must be > 0, got $a4Hz" }
        val midiFloat = 69f + 12f * (ln(freqHz / a4Hz) / LN2).toFloat()
        val midi = round(midiFloat).toInt()
        val cents = (midiFloat - midi) * 100f
        return PitchEstimate(freqHz, midi.coerceIn(0, 127), cents)
    }

    /** Inverse of [analyze]: equal-tempered frequency for the given MIDI note. */
    fun midiToFreq(midi: Int, a4Hz: Float = 440f): Float =
        a4Hz * 2f.pow((midi - 69) / 12f)

    private const val LN2 = 0.6931471805599453
}

File: chorect-web/src/audio/pitchAnalysis.ts
// Equal-tempered pitch math with configurable A4, ported from audio/.../PitchAnalysis.kt.

export interface PitchEstimate {
  /** Detected fundamental in Hz. */
  readonly freqHz: number;
  /** Nearest equal-tempered MIDI note, clamped to 0..127. */
  readonly midi: number;
  /** Signed cents offset from the nearest note, in (-50, +50]. */
  readonly cents: number;
}

const LN2 = 0.6931471805599453;

export function analyzePitch(freqHz: number, a4Hz = 440): PitchEstimate {
  if (freqHz <= 0) throw new Error(`freq must be > 0, got ${freqHz}`);
  if (a4Hz <= 0) throw new Error(`A4 reference must be > 0, got ${a4Hz}`);
  const midiFloat = 69 + (12 * Math.log(freqHz / a4Hz)) / LN2;
  const midi = Math.round(midiFloat);
  const cents = (midiFloat - midi) * 100;
  return { freqHz, midi: Math.min(Math.max(midi, 0), 127), cents };
}

export function midiToFreqA4(midi: number, a4Hz = 440): number {
  return a4Hz * Math.pow(2, (midi - 69) / 12);
}
